vllm/v1/engine/core.py

```diff
@@ -205,23 +205,18 @@ def step_with_batch_queue(self) -> Optional[EngineCoreOutputs]:
                 self.batch_queue.put_nowait(
                     (future, scheduler_output))  # type: ignore
 
-        # If all requests are scheduled or the job queue is full,
+        scheduled_batch = (scheduler_output is not None
+                           and scheduler_output.total_num_scheduled_tokens > 0)
+
+        # If no more requests can be scheduled and the job queue is not empty,
         # block until the first batch in the job queue is finished.
-        if (scheduler_output is None
-                or scheduler_output.total_num_scheduled_tokens == 0):
-            try:
-                future, scheduler_output = self.batch_queue.get(
-                    timeout=POLLING_TIMEOUT_S)
-                # Blocking until the first result is available.
-                model_output = future.result()
-                self.batch_queue.task_done()
-                engine_core_outputs = self.scheduler.update_from_output(
-                    scheduler_output, model_output)
-            except queue.Empty:
-                # If the queue is empty (timeout at .get), return
-                # an empty EngineCoreOutputs for logging.
-                engine_core_outputs = EngineCoreOutputs(
-                    outputs=[], scheduler_stats=self.scheduler.make_stats())
+        if not scheduled_batch and not self.batch_queue.empty():
+            future, scheduler_output = self.batch_queue.get_nowait()
+            # Blocking until the first result is available.
+            model_output = future.result()
```

Comment:
We don't need any timeout here?

Reply:
No because in this case we have nothing else to do but can just wait for the first batch to finish (similar to PP=1)

Reply:
but in this case, the engine will not add any requests to the scheduler in the meantime, right?

Reply:
Yes but doesn't matter because all GPUs are busy. On the other hand if we no wait here and add new requests to the scheduler, we will suffer from two issues:
1. The new added requests cannot be scheduled anyways due to no resources.
2. If the oldest batch is finished during the time we are adding new requests, then that batch will be delayed.

Reply:
I see. Thanks for the explanation!

Reply:
oh btw what about the FSM compilation? Can it be done in parallel with the execution?

Reply:
I guess so, although we won't be able to schedule the request before its FSM is compiled, we should be able to kick off the compilation first.